apps/showcase/src/services/webcontainer/webcontainer-runner.ts

```diff
@@ -0,0 +1,275 @@
+import {FileSystemTree, IFSWatcher, WebContainer, WebContainerProcess} from '@webcontainer/api';
+import {Terminal} from '@xterm/xterm';
+import {
+  BehaviorSubject,
+  combineLatestWith,
+  distinctUntilChanged,
+  filter,
+  from,
+  map,
+  Observable,
+  switchMap
+} from 'rxjs';
+
+const createTerminalStream = (terminal: Terminal, cb?: (data: string) => void | Promise<void>) => new WritableStream({
+  write: (data) => {
+    if (cb) {
+      void cb(data);
+    }
+    terminal.write(data);
+  }
+});
+
+const makeProcessWritable = (process: WebContainerProcess, terminal: Terminal) => {
+  const input = process.input.getWriter();
+  terminal.onData((data) => input.write(data));
+  return input;
+};
+
+export class WebContainerRunner {
+  /**
+   * WebContainer instance which is available after the
+   */
+  public readonly instance: Promise<WebContainer>;
+  private readonly commands = new BehaviorSubject<{queue: string[]; cwd: string}>({queue: [], cwd: ''});
+  private readonly commandOnRun: Observable<{command: string; cwd: string} | undefined> = this.commands.pipe(
+    map((commands) => (
+      commands.queue.length > 0 ? {command: commands.queue[0], cwd: commands.cwd} : undefined
+    ))
+  );
+  private readonly iframe = new BehaviorSubject<HTMLIFrameElement | null>(null);
+  private readonly shell = {
+    terminal: new BehaviorSubject<Terminal | null>(null),
+    process: new BehaviorSubject<WebContainerProcess | null>(null),
+    writer: new BehaviorSubject<WritableStreamDefaultWriter | null>(null),
+    cwd: new BehaviorSubject<string | null>(null)
+  };
+  private readonly commandOutput = {
+    terminal: new BehaviorSubject<Terminal | null>(null),
+    process: new BehaviorSubject<WebContainerProcess | null>(null),
+    outputLocked: new BehaviorSubject<boolean>(false)
+  };
+  private watcher: IFSWatcher | null = null;
+
+  constructor() {
+    this.instance = new Promise((resolve) => {
+      void WebContainer.boot().then((instance) => {
+        resolve(instance);
+        // eslint-disable-next-line no-console
+        instance.on('error', console.error);
+      });
+    });
+    this.commandOnRun.pipe(
+      filter((currentCommand): currentCommand is {command: string; cwd: string} => !!currentCommand)
```

Comment:
```suggestion
      filter((currentCommand): currentCommand is {command: string; cwd: string} => !!currentCommand),
      takeUntilDestroyed()
```

from `@angular/core/rxjs-interop`

Reply:
Yes, I forgot to clean the subscriptions and it is no longer a service (which is something we may discuss because WebContainer.boot() only can run once)